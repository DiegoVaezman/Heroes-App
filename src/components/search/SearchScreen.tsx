import React, { useEffect, useMemo } from "react";
import { useForm } from "../../hooks/useForm";
import { getHeroByName } from "../../selectors/getHeroByName";
import { HeroCard } from "../hero/HeroCard";
import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const useQuery = () => {
    return new URLSearchParams(location.search);
  };
  const query = useQuery().get("q") || "";
  const { onChange, searchText } = useForm({
    searchText: query,
  });

  const heroesFiltered = useMemo(() => getHeroByName(query), [query]);

  const handleSearch = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Búsquedas</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Buscar</h4>
          <hr />

          <form onSubmit={(e) => handleSearch(e)}>
            <input
              type="text"
              placeholder="Buscar un héroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={(event) => onChange(event.target.value, "searchText")}
            />
            <button type="submit" className="btn btn-outline-primary mt-1">
              Buscar...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />
          {query === "" ? (
            <div className="alert alert-info">Buscar un héroe</div>
          ) : (
            heroesFiltered.length === 0 && (
              <div className="alert alert-info">No hay resultados: {query}</div>
            )
          )}
          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} hero={hero} />
          ))}
        </div>
      </div>
    </>
  );
};
