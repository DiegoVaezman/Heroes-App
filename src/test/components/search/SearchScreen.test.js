import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
}));
const mockUseNavigate = jest.fn();

describe('Testing SearchScreen', () => {
    //LIMPIAR LOS MOCKS DESPUÃ‰S DE CADA TEST PARA QUE NO INTERFIERAN
    beforeEach(() => jest.clearAllMocks());

    test('should show default values', () => {
        const { container } = render(
            <MemoryRouter>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });

    test('should show batman and input with query value', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        const img = screen.getByRole('img');
        expect(img.src).toContain('/assets/dc-batman.jpg');

        expect(screen.queryByText('Search an hero')).toBeNull();
    });

    test('should show an error if hero is not found', () => {
        const value = 'batman123';
        render(
            <MemoryRouter initialEntries={[`/search?q=${value}`]}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(screen.getByText(`No hero with: ${value}`)).toBeTruthy();
    });

    test('should call navigate to new screen', () => {
        const value = 'superman';
        render(
            <MemoryRouter>
                <SearchScreen />
            </MemoryRouter>
        );
        const input = screen.getByRole('textbox');
        fireEvent.change(input, {
            target: { name: 'searchText', value: 'superman' },
        });
        screen.debug();

        //el form no se pilla con getByRole a no ser que se le indique en el componente role='form'
        const form = screen.getByRole('form');
        fireEvent.submit(form);
        // const searchButton = screen.getByRole('button', { name: 'Search...' });
        // fireEvent.click(searchButton);
        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${value}`);
    });
});
