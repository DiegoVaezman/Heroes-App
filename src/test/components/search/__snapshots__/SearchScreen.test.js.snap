// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchScreen should show default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        role="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info animate__animated animate__fadeIn"
      >
        Search an hero
      </div>
    </div>
  </div>
</div>
`;
